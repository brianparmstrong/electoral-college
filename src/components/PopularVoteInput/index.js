import React, { useState } from 'react';
import { arrayOf, func, number, oneOfType, string } from 'prop-types';

const PopularVoteInput = ({
  currentPVTotals,
  evs,
  handlePropVotes,
  name,
  party,
  percent,
  stateEvs
}) => {
  const initialInputValue = percent ? percent : '';
  const [evsAwarded, setEvsAwarded] = useState(0);
  const [inputValue, setInputValue] = useState(initialInputValue);
  const [savedValue, setSavedValue] = useState('');
  const inputId = party + 'PopVoteInput';
  const inputPlaceholder = `${party.toUpperCase()} %`;

  const calculateProportionalVotes = (e, value) => {
    const newInputValue = value ? Number(value) : (e && e.target) ? Number(e.target.value) : 0;
    let totalEvs = stateEvs ? stateEvs : evs;
    const popVotePercent = (Boolean(newInputValue) || percent) ? (newInputValue / 100) : 0;
    const evsWon = Math.round(totalEvs * popVotePercent);
    const polParty = (value || Boolean(e) === false) ? null : e.target.getAttribute('data-party');
    let newGopPropTotal = currentPVTotals[0];
    let newDemPropTotal = currentPVTotals[1];
    let newLibPropTotal = currentPVTotals[2];
    let newGrnPropTotal = currentPVTotals[3];
    let newIndPropTotal = currentPVTotals[4];

    if (Boolean(evsWon) && (newInputValue <= 100)) {
      switch (polParty) {
        case 'gop':
          newGopPropTotal += evsWon;
          break;
        case 'dem':
          newDemPropTotal += evsWon;
          break;
        case 'lib':
          newLibPropTotal += evsWon;
          break;
        case 'grn':
          newGrnPropTotal += evsWon;
          break;
        case 'ind':
          newIndPropTotal += evsWon;
          break;
        default:
          break;
      }

      const newPVTotals = [
        newGopPropTotal,
        newDemPropTotal,
        newLibPropTotal,
        newGrnPropTotal,
        newIndPropTotal
      ];

      if (e && e.target) {
        e.target.setAttribute('data-evsawarded', evsWon);
        handlePropVotes(newPVTotals);
      } else if (percent) {
        return evsWon;
      }
    }
  };

  const resetEvsWon = (e, newEvsAwarded) => {
    const politicalParty = e.target.getAttribute('data-party');
    let newGopPropTotal = currentPVTotals[0];
    let newDemPropTotal = currentPVTotals[1];
    let newLibPropTotal = currentPVTotals[2];
    let newGrnPropTotal = currentPVTotals[3];
    let newIndPropTotal = currentPVTotals[4];

    switch (politicalParty) {
      case 'gop':
        newGopPropTotal -= newEvsAwarded;
        break;
      case 'dem':
        newDemPropTotal -= newEvsAwarded;
        break;
      case 'lib':
        newLibPropTotal -= newEvsAwarded;
        break;
      case 'grn':
        newGrnPropTotal -= newEvsAwarded;
        break;
      case 'ind':
        newIndPropTotal -= newEvsAwarded;
        break;
      default:
        break;
    }

    const newPVTotals = [
        newGopPropTotal,
        newDemPropTotal,
        newLibPropTotal,
        newGrnPropTotal,
        newIndPropTotal
      ];

    e.target.setAttribute('data-evsawarded', 0);
    handlePropVotes(newPVTotals);
  };

  const handleInputBlur = e => {
    e.persist();
    if (inputValue !== '' && e.target.value === '') {
      setInputValue(savedValue);
      calculateProportionalVotes(null, savedValue);
    } else {
      calculateProportionalVotes(e);
    }
  };

  const handleInputFocus = e => {
    const evsAwardedData = Number(e.target.getAttribute('data-evsawarded'));
    const newEvsAwarded = isNaN(evsAwardedData) ? evsAwarded : parseInt(evsAwardedData);
    e.persist();
    setEvsAwarded(newEvsAwarded);
    setSavedValue(inputValue);
    setInputValue('');
    resetEvsWon(e, newEvsAwarded);
  };

  return (
    <div className='pvInputWrapper'>
      <input
        type='text'
        className='pvInput'
        id={inputId}
        pattern="^(100(?:\.00)?|0(?:\.\d\d)?|\d?\d(?:\.\d\d)?)$"
        placeholder={inputPlaceholder}
        data-evsawarded={String(calculateProportionalVotes(null, inputValue))}
        data-party={party}
        data-statename={name}
        defaultValue={inputValue}
        onBlur={handleInputBlur}
        onFocus={handleInputFocus} />
    </div>
  );
};

PopularVoteInput.propTypes = {
  currentPVTotals: arrayOf(number).isRequired,
  evs: oneOfType([number, string]).isRequired,
  handlePropVotes: func.isRequired,
  name: string.isRequired,
  party: string.isRequired,
  percent: oneOfType([number, string]).isRequired,
  stateEvs: number
};

export default PopularVoteInput;
