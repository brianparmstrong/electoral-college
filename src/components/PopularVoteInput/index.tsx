import { FocusEvent, useState } from 'react';
import { IfcPopularVoteInput } from '../../types';

const PopularVoteInput = (popularVoteInput: IfcPopularVoteInput) => {
  const {
    currentPVTotals,
    evs,
    handlePropVotes,
    name,
    party,
    percent,
    stateEvs,
  } = popularVoteInput;
  const initialInputValue = percent ? percent : '';
  const [evsAwarded, setEvsAwarded] = useState(0);
  const [inputValue, setInputValue] = useState(initialInputValue);
  const [savedValue, setSavedValue] = useState('');
  const inputId = `${party}PopVoteInput`;
  const inputPlaceholder = `${party.toUpperCase()} %`;

  const calculateProportionalVotes = (
    event: FocusEvent<HTMLInputElement>,
    savedValue?: string
  ) => {
    const eventTarget = event?.target as HTMLInputElement;
    const value = eventTarget?.value;
    /* eslint-disable indent */
    const newInputValue = savedValue
      ? Number(savedValue)
      : eventTarget
      ? Number(value)
      : 0;
    /* eslint-enable indent */
    let totalEvs = stateEvs ?? evs;
    if (name === 'Maine') {
      totalEvs = '4';
    } else if (name === 'Nebraska') {
      totalEvs = '5';
    }
    const popVotePercent =
      Boolean(newInputValue) || percent ? newInputValue / 100 : 0;
    const evsWon = Math.round(Number(totalEvs) * popVotePercent);
    const polParty =
      savedValue || Boolean(event) === false
        ? null
        : event.target.dataset.party;
    let newGopPropTotal = currentPVTotals[0];
    let newDemPropTotal = currentPVTotals[1];
    let newLibPropTotal = currentPVTotals[2];
    let newGrnPropTotal = currentPVTotals[3];
    let newIndPropTotal = currentPVTotals[4];

    if (Boolean(evsWon) && newInputValue <= 100) {
      switch (polParty) {
        case 'gop':
          newGopPropTotal += evsWon;
          break;
        case 'dem':
          newDemPropTotal += evsWon;
          break;
        case 'lib':
          newLibPropTotal += evsWon;
          break;
        case 'grn':
          newGrnPropTotal += evsWon;
          break;
        case 'ind':
          newIndPropTotal += evsWon;
          break;
        default:
          break;
      }

      const newPVTotals = [
        newGopPropTotal,
        newDemPropTotal,
        newLibPropTotal,
        newGrnPropTotal,
        newIndPropTotal,
      ];

      if (event?.target) {
        event.target.setAttribute('data-evsawarded', String(evsWon));
        handlePropVotes(newPVTotals);
      } else if (percent) {
        return evsWon;
      }
    }
  };

  const resetEvsWon = (
    event: FocusEvent<HTMLInputElement>,
    newEvsAwarded: number
  ) => {
    const politicalParty = event.currentTarget.dataset.party;
    let newGopPropTotal = currentPVTotals[0];
    let newDemPropTotal = currentPVTotals[1];
    let newLibPropTotal = currentPVTotals[2];
    let newGrnPropTotal = currentPVTotals[3];
    let newIndPropTotal = currentPVTotals[4];

    switch (politicalParty) {
      case 'gop':
        newGopPropTotal -= newEvsAwarded;
        break;
      case 'dem':
        newDemPropTotal -= newEvsAwarded;
        break;
      case 'lib':
        newLibPropTotal -= newEvsAwarded;
        break;
      case 'grn':
        newGrnPropTotal -= newEvsAwarded;
        break;
      case 'ind':
        newIndPropTotal -= newEvsAwarded;
        break;
      default:
        break;
    }

    const newPVTotals = [
      newGopPropTotal,
      newDemPropTotal,
      newLibPropTotal,
      newGrnPropTotal,
      newIndPropTotal,
    ];

    event.target.setAttribute('data-evsawarded', '0');
    handlePropVotes(newPVTotals);
  };

  const handleInputBlur = (event: FocusEvent<HTMLInputElement>) => {
    event.persist();
    const { value } = event.currentTarget as HTMLInputElement;
    if (inputValue !== '' && value === '') {
      setInputValue(savedValue);
      calculateProportionalVotes(null, savedValue);
    } else {
      calculateProportionalVotes(event);
    }
  };

  const handleInputFocus = (event: FocusEvent<HTMLInputElement>) => {
    const evsAwardedData = Number(event.currentTarget.dataset.evsawarded);
    const newEvsAwarded = isNaN(evsAwardedData) ? evsAwarded : evsAwardedData;
    event.persist();
    setEvsAwarded(newEvsAwarded);
    setSavedValue(inputValue);
    setInputValue('');
    resetEvsWon(event, newEvsAwarded);
  };

  return (
    <div className="pvInputWrapper">
      <input
        type="text"
        className="pvInput"
        id={inputId}
        pattern="^(100(?:\.00)?|0(?:\.\d\d)?|\d?\d(?:\.\d\d)?)$"
        placeholder={inputPlaceholder}
        data-evsawarded={String(calculateProportionalVotes(null, inputValue))}
        data-party={party}
        data-statename={name}
        defaultValue={inputValue}
        onBlur={handleInputBlur}
        onFocus={handleInputFocus}
      />
    </div>
  );
};

export default PopularVoteInput;
