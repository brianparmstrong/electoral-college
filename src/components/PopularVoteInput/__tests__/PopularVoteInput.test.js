import { render } from '@testing-library/react';
import PopularVoteInput from '../';

const MOCK_PV_INPUT_PROPS = {
  currentPVTotals: [100, 90, 5, 2, 1],
  evs: '20',
  handlePropVotes: jest.fn(),
  name: 'Pennsylvania',
  party: 'grn',
  percent: '10',
};

it('renders correctly', () => {
  const { container } = render(<PopularVoteInput {...MOCK_PV_INPUT_PROPS} />);
  expect(container.firstChild).toMatchSnapshot();
});
