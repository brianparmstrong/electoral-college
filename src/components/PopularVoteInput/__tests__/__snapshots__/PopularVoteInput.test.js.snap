// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="pvInputWrapper"
>
  <input
    class="pvInput"
    data-evsawarded="2"
    data-party="grn"
    data-statename="Pennsylvania"
    id="grnPopVoteInput"
    pattern="^(100(?:\\\\.00)?|0(?:\\\\.\\\\d\\\\d)?|\\\\d?\\\\d(?:\\\\.\\\\d\\\\d)?)$"
    placeholder="GRN %"
    type="text"
    value="10"
  />
</div>
`;
